textx/registration.py

```diff
@@ -0,0 +1,312 @@
+    """
+    global languages
+    if languages is None:
+        languages = {}
+        for language in pkg_resources.iter_entry_points(
+                group='textx_languages'):
+            register_language(language.load())
+    return languages
+
+
+def generator_descriptions():
+    """
+    Return a dict of `GeneratorDesc` instances keyed by language name.
+    """
+    global generators
+    if generators is None:
+        generators = {}
+        for generator in pkg_resources.iter_entry_points(
+                group='textx_generators'):
+            register_generator(generator.load())
+    return generators
+
+
+def language_description(language_name):
+    """
+    Return `LanguageDesc` for the given language name.
+    """
+    global languages
+    language_name = language_name.lower()
+    if languages is None:
+        language_descriptions()
+    try:
+        return languages[language_name]
+    except KeyError:
+        raise TextXRegistrationError('Language "{}" not registered.'
+                                     .format(language_name))
+
+
+def generator_description(language_name, target_name, any_permitted=False):
+    """
+    Return `GeneratorDesc` instance for the given target and language name.
+    If `any_permitted` is `True` return generator for language `any` if
+    generator for the `language_name` doesn't exist.
+    """
+    global generators
+    language_name = language_name.lower()
+    target_name = target_name.lower()
+    if generators is None:
+        generator_descriptions()
+    try:
+        try:
+            generators_for_language = generators[language_name]
+            return generators_for_language[target_name]
+        except KeyError:
+            if not any_permitted:
+                raise
+            generators_for_language = generators['any']
+            return generators_for_language[target_name]
+    except KeyError:
+        raise TextXRegistrationError(
+            'No generators registered for language "{}" and target "{}".'
+            .format(language_name, target_name))
+
+
+def generator_for_language_target(language_name, target_name,
+                                  any_permitted=False):
+    """
+    Return generator callable for the given language name and target name.
+    """
+    generator_desc = generator_description(language_name, target_name,
+                                           any_permitted=any_permitted)
+    return generator_desc.generator
+
+
+def register_language(language_desc_or_name, pattern=None, description='',
+                      metamodel=None):
+    """
+    Programmatically register a language.
+
+    Args:
+        language_desc_or_name (LanguageDesc or str): If LanguageDesc is given
+            other parameters are not used.
+        For other parameters see `LanguageDesc`.
+    """
+    global languages
+    if languages is None:
+        language_descriptions()
+
+    if type(language_desc_or_name) is not LanguageDesc:
```

Comment:
@igordejanovic Can you change this to check if `language_desc_or_name` is instance of `LanguageDesc`?
The reason is that we will need additional fields/methods for language description for `textX-LS`, and I thought it would be a good idea to avoid replacing similar logic in `textX-LS-core` and to allow `textX` to handle subclasses of `LanguageDesc`.

Reply:
Sure. It sounds like a good idea.

Reply:
Done, for both languages and generators.